lib/array.cpp

```diff
@@ -404,20 +408,34 @@ TYPE scalar::cast() const
   values_holder v;
   inject(v);
 
-#define HANDLE_CASE(DTYPE, VAL) case DTYPE: return static_cast<TYPE>(v.VAL)
+//#define HANDLE_CASE(DTYPE, VAL) case DTYPE: return static_cast<TYPE>(v.VAL)
+//#define HANDLE_CASE(FLOAT_TYPE) case
+
 
   switch(dtype_)
   {
-    HANDLE_CASE(CHAR_TYPE, int8);
-    HANDLE_CASE(UCHAR_TYPE, uint8);
-    HANDLE_CASE(SHORT_TYPE, int16);
-    HANDLE_CASE(USHORT_TYPE, uint16);
-    HANDLE_CASE(INT_TYPE, int32);
-    HANDLE_CASE(UINT_TYPE, uint32);
-    HANDLE_CASE(LONG_TYPE, int64);
-    HANDLE_CASE(ULONG_TYPE, uint64);
-    HANDLE_CASE(FLOAT_TYPE, float32);
-    HANDLE_CASE(DOUBLE_TYPE, float64);
+//    HANDLE_CASE(CHAR_TYPE, int8);
+//    HANDLE_CASE(UCHAR_TYPE, uint8);
+//    HANDLE_CASE(SHORT_TYPE, int16);
+//    HANDLE_CASE(USHORT_TYPE, uint16);
+//    HANDLE_CASE(INT_TYPE, int32);
+//    HANDLE_CASE(UINT_TYPE, uint32);
+//    HANDLE_CASE(LONG_TYPE, int64);
+//    HANDLE_CASE(ULONG_TYPE, uint64);
+//    HANDLE_CASE(FLOAT_TYPE, float32);
+//    HANDLE_CASE(DOUBLE_TYPE, float64);
+//    HANDLE_CASE(HALF_TYPE, float16);
+    case CHAR_TYPE:    { char t = v.int8;return t;} break;
+    case UCHAR_TYPE: { unsigned char t = v.uint8;return t;} break;
+    case SHORT_TYPE:{ short t = v.int16;return t;} break;
+    case USHORT_TYPE:{ unsigned short t = v.uint16;return t;} break;
+    case INT_TYPE:{ int t = v.int32;return t;} break;
+    case UINT_TYPE:{ unsigned int t = v.uint32;return t;} break;
+    case LONG_TYPE: {long t = v.int64;return t;} break;
+    case ULONG_TYPE:{ unsigned long t =  v.uint64;return t;} break;
+    case HALF_TYPE: (half)(v.float16);
+    case FLOAT_TYPE:{float t = v.float32;return t;} break;
+    case DOUBLE_TYPE: { double t = v.float64;return t;} break;
     default: throw unknown_datatype(dtype_);
   }
```

Comment:
Wouldn't it work to just keep a macro here (if the union is not changed to a struct), even for the half-type?